---
title: "R Review Project"
output: html_notebook
editor_options: 
  chunk_output_type: console
---


Directory set up, package download, and library loading
```{r}
#set directory
knitr::opts_knit$set(root.dir = normalizePath("/Users/andreavelazquez/Desktop/School stuff/QBIO490/qbio_490_AndreaV/analysis_data")) 


#Check if packages are ready to be loaded, if not download them
if(!require(BiocManager)) {
    install.packages("BiocManager")
}

if(!require(TCGAbiolinks)) {
    install.packages("TCGAbiolinks")
}

if(!require(maftools)) {
    install.packages("maftools")
}

if (!require(survival)) {
  install.packages("survival")
}

if (!require(survminer)) {
  install.packages("survminer")
}

if (!require(ggplot2)) {
  install.packages("ggplot2")
}

if (!require(DESeq2)) {
  install.packages("DESeq2")
}

if (!require(SummarizedExperiment)) {
  install.packages("SummarizedExperiment")
}

if (!require(EnhancedVolcano)) {
  install.packages("EnhancedVolcano")
}




#Load packages
library(BiocManager)
library(TCGAbiolinks)
library(maftools)
library(survival)
library(survminer)
library(ggplot2)
library(DESeq2)
library(SummarizedExperiment)
library(EnhancedVolcano)


```

Querying data
```{r}
#===============================================================================
#Query clinical data
clin_query <- GDCquery(project = "TCGA-SKCM",
              data.category = "Clinical",
              data.type = "Clinical Supplement",
              data.format = 'BCR Biotab')

#GDCdownload(clin_query)
clinical.BCRtab.all <- GDCprepare(clin_query)
  
#clinic <- clinical.BCRtab.all$clinical_patient_brca[-c(1,2),]
rad<- clinical.BCRtab.all$clinical_radiation_skcm[-c(1, 2), ]
drugs<- clinical.BCRtab.all$clinical_drug_skcm[-c(1, 2), ]

#===============================================================================
#Query rna data
rna_query <- GDCquery(project ="TCGA-SKCM",
            data.category = "Transcriptome Profiling",
            data.type = "Gene Expression Quantification",
            workflow.type = "STAR - Counts")

#Download. Comment out after first time
#GDCdownload(rna_query)
rna_se <- GDCprepare(rna_query)
#make df from rna_se object w clinical data
rna_clinical<- rna_se@colData
rna_clinical<- as.data.frame(rna_clinical)
colnames(rna_clinical)[ colnames(rna_clinical) == "patient" ] <- "Tumor_Sample_Barcode"

#===============================================================================
#Query maf data
maf_query <- GDCquery(project = "TCGA-SKCM",
                      data.category = "Simple Nucleotide Variation",
                      access = "open",
                      data.type = "Masked Somatic Mutation",
                      workflow.type = "Aliquot Ensemble Somatic Variant Merging and Masking")

#Download. Comment out after first time
#GDCdownload(maf_query)
maf <- GDCprepare(maf_query)
SKCM_maf <- read.maf(maf = maf,
                     clinicalData = rna_clinical,
                     isTCGA = TRUE)

#===============================================================================
```

Make data easier to see and manipulate
```{r}

#rna_genes df
rna_genes<- rna_se@rowRanges@elementMetadata
rna_genes<- as.data.frame(rna_genes)

#rna_counts df
rna_counts<- rna_se@assays@data$unstranded
rna_counts<- as.data.frame(rna_counts)

#change row and column names for easier understadning
rownames(rna_genes)<- rna_genes$gene_id
rownames(rna_counts)<- rna_genes$gene_id
colnames(rna_counts)<- rownames(rna_clinical)
```


1. Evaluating the difference in survival between metastatic and non-metastatic SKCM patients
```{r}

#No need to maks for NAs when looking at the patients' metastatic status becasue there is no NA values
#checked w unique(rna_se@colData$definition)
#Make new column to identify metastatic patients
PST_M_mask<- ifelse(rna_clinical$definition == "Metastatic", 
                    T,
                    ifelse( rna_clinical$definition== "Primary solid Tumor", 
                            T, 
                            F))

isMetastatic<- ifelse(rna_clinical$definition == "Metastatic", 
                      T, 
                      F)

rna_clinical<- rna_clinical[PST_M_mask, ]
rna_counts<- rna_counts[ , PST_M_mask]
rna_clinical$isMetastatic<- isMetastatic



#Make new column to have the survival time
#If patient is not dead, use the days to last follow up
rna_clinical$survival_time<- ifelse(is.na(rna_clinical$days_to_death), 
                                    rna_clinical$days_to_last_follow_up, 
                                    rna_clinical$days_to_death)

rna_clinical$survival_time<- as.numeric(rna_clinical$survival_time)



#Make new column for death event
rna_clinical$death_event<- ifelse(rna_clinical$vital_status== "Dead", T, F)



#Create survival object
survival_object <- Surv(time = rna_clinical$survival_time,
                        event = rna_clinical$death_event)

fit_object <- survfit(survival_object ~ isMetastatic, data = rna_clinical)



#Plot
survplot <- ggsurvplot(fit_object,
                       pval=TRUE,
                       ggtheme = theme(plot.margin = unit(c(1,1,1,1), "cm")),
                       legend = 'right', 
                       legend.labs= c("Non-metastatic patients", "Metastatic patients"))

KM_plot <- survplot$plot + theme_bw() + theme(axis.title = element_text(size=20), 
                                              axis.text = element_text(size=16),
                                              legend.title = element_text(size=14),
                                              legend.text = element_text(size=12))

KM_plot

```

2. Evaluating mutation differences for metastatic and non-metastatic patients
```{r}
#Make factor 
rna_clinical$isMetastatic<- factor(rna_clinical$isMetastatic)

#Get barcodes for subsetting
isMetastatic_barcodes<- rna_clinical$Tumor_Sample_Barcode[isMetastatic]
notMetastatic_barcodes<- rna_clinical$Tumor_Sample_Barcode[!isMetastatic]

#subset
maf_isM<- subsetMaf(maf= SKCM_maf,
                    tsb= isMetastatic_barcodes)

maf_notM<- subsetMaf(maf= SKCM_maf,
                    tsb= notMetastatic_barcodes)

#Get list of top 10 genes
m1.genes = getGeneSummary(maf_isM)[1:10]
m2.genes = getGeneSummary(maf_notM)[1:10]

mdt = merge(m1.genes[,.(Hugo_Symbol, MutatedSamples)], 
            m2.genes[,.(Hugo_Symbol, MutatedSamples)], 
            by = 'Hugo_Symbol', all = TRUE)

mdt$MutatedSamples.x[is.na(mdt$MutatedSamples.x)] = 0
mdt$MutatedSamples.y[is.na(mdt$MutatedSamples.y)] = 0

mdt$max = apply(mdt[,.(MutatedSamples.x, 
                       MutatedSamples.y)], 
                1, 
                max)
mdt = mdt[order(max, decreasing = TRUE)]

#plot
coOncoplot(m1=maf_isM, 
           m2=maf_notM, 
           m1Name = "Metastatic Patients",
           m2Name = "Non-metastatic Patients",
           genes= mdt$Hugo_Symbol,
           borderCol = NA)

```

3. Compare mutation types and loci for the FAT4 gene across metastatic and non-metastatic patients
```{r}

lollipopPlot2(m1= maf_isM, 
              m2= maf_notM,
              m1_name = "Metastatic SKCM", 
              m2_name = "Non-metastatic SKCM",
              gene="FAT4",
              showDomainLabel = TRUE,
              pointSize = c(1.3))#made the lollipop heads a bit smaller so it's not too visually overwhelming but they're still visible

```

4. Coocurrence or mutualexclusion of common gene mutations. One for metastatic and one for non-metastatic patients
```{r}

somaticInteractions(maf_notM, top=25, pvalue= c(0.05, 0.1), pvSymbols = c(23, 42))

somaticInteractions(maf_isM, top=25, pvalue= c(0.05, 0.1), pvSymbols = c(23, 42))

```

5. Differential expression for metastatic vs non-metastatic patients
First, perform one-hot encoding for select treatments, as mentioned in the instructions for this project.
```{r}
#One-hot encoding for radiation, chemotherapy, immunotherapy, molecular therapy, and vaccine treatments
#===============================================================================
#Radiation
rna_clinical$radiation <- ifelse(rna_clinical$Tumor_Sample_Barcode %in% rad$bcr_patient_barcode, 
                                 1, 
                                 0) 

#===============================================================================
#Chemotherapy
had_chemo<- ifelse(drugs$pharmaceutical_therapy_type == "Chemotherapy", 
                   T, 
                   F)

#New temp df for comparison in encoding
chemo_pt<- drugs[had_chemo, ]
#1 if pt is in chemo_pt df
rna_clinical$chemotherapy<- ifelse(rna_clinical$Tumor_Sample_Barcode %in% chemo_pt$bcr_patient_barcode, 
                                   1, 
                                   0)

#===============================================================================
#Immunotherapy
had_immunotherapy<- ifelse(drugs$pharmaceutical_therapy_type == "Immunotherapy", 
                   T, 
                   F)

#New temp df for comparison in encoding
immunotherapy_pt<- drugs[had_immunotherapy, ]
#1 if pt is in chemo_pt df
rna_clinical$immunotherapy<- ifelse(rna_clinical$Tumor_Sample_Barcode %in% immunotherapy_pt$bcr_patient_barcode, 
                                   1, 
                                   0)

#===============================================================================
#Molecular therapy
had_mol_therapy<- ifelse(drugs$pharmaceutical_therapy_type == "Targeted Molecular therapy", 
                   T, 
                   F)

#New temp df for comparison in encoding
mol_therapy_pt<- drugs[had_mol_therapy, ]
#1 if pt is in chemo_pt df
rna_clinical$molecular_therapy<- ifelse(rna_clinical$Tumor_Sample_Barcode %in% mol_therapy_pt$bcr_patient_barcode, 
                                   1, 
                                   0)

#===============================================================================
#Vaccine
had_vaccine<- ifelse(drugs$pharmaceutical_therapy_type == "Vaccine", 
                   T, 
                   F)

#New temp df for comparison in encoding
vaccine_pt<- drugs[had_vaccine, ]
#1 if pt is in chemo_pt df
rna_clinical$vaccine<- ifelse(rna_clinical$Tumor_Sample_Barcode %in% vaccine_pt$bcr_patient_barcode, 
                                   1, 
                                   0)

#===============================================================================
#Turn all columns into factors for DESeq2
rna_clinical$radiation<- factor(rna_clinical$radiation)
rna_clinical$chemotherapy<- factor(rna_clinical$chemotherapy)
rna_clinical$immunotherapy<- factor(rna_clinical$immunotherapy)
rna_clinical$molecular_therapy<- factor(rna_clinical$molecular_therapy)
rna_clinical$vaccine<- factor(rna_clinical$vaccine)

```

Some more cleaning up to run DESeq2 after
```{r}
#Vital status, gender, race = covariates
#no need to cleanup for NA, unique() for each and no NA!
#===============================================================================
#Vital status
vital_status_NR_mask<- ifelse(rna_clinical$vital_status != "Not Reported", 
                              T, 
                              F)
rna_clinical<- rna_clinical[vital_status_NR_mask, ]
rna_counts<- rna_counts[ , vital_status_NR_mask]

#===============================================================================
race_NR_mask<- ifelse(rna_clinical$race != "not reported", 
                      T, 
                      F)

rna_clinical<- rna_clinical[race_NR_mask, ]
rna_counts<- rna_counts[ , race_NR_mask]

#===============================================================================
#No need to do gender, 0 not reported
#Turn all of the covariates into factors
rna_clinical$gender<- factor(rna_clinical$gender)
rna_clinical$vital_status<- factor(rna_clinical$vital_status)
rna_clinical$race<- factor(rna_clinical$race)

#===============================================================================
#Filter out genes w counts <20 
lowcount_mask<- ifelse(rowSums(rna_counts) < 20, F, T)
rna_counts<- rna_counts[lowcount_mask, ]
rna_genes<- rna_genes[lowcount_mask, ]

```

Set up DESeq2 analysis
```{r}
#make dataset
SKCM_dataset <- DESeqDataSetFromMatrix(countData = rna_counts,
                                       colData = rna_clinical,
                                       design= ~ race + gender + vital_status + radiation + chemotherapy + immunotherapy + molecular_therapy + vaccine + isMetastatic)

#make object
SKCM_DSEQ_object<- DESeq(SKCM_dataset)

#extract results
results <- results(SKCM_DSEQ_object, format = "DataFrame", 
                   contrast = c("isMetastatic", "TRUE", "FALSE"))

EnhancedVolcano(results,
                lab = rownames(results),
                pCutoff= 0.05, 
                FCcutoff= 1,
                x = 'log2FoldChange',
                y = 'pvalue', 
                ylab= "-log10(P_Value)", 
                xlab= "log2(Fold_Change)", 
                title = "Gene differential expression for metastatics vs non-metastatic patients")


```




