{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0ae81e67",
   "metadata": {},
   "source": [
    "# Python Review Project"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "id": "342ee7c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#SET WD\n",
    "import os\n",
    "os.chdir('/Users/andreavelazquez/Desktop/School stuff/QBIO490/qbio_490_AndreaV/analysis_data')\n",
    "\n",
    "import cptac\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.preprocessing import LabelEncoder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "id": "90f0790d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                          \r"
     ]
    }
   ],
   "source": [
    "#IMPORT CPTAC\n",
    "cptac.download(dataset=\"Ccrcc\")\n",
    "ccrcc= cptac.Ccrcc()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "id": "4ba0726e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#RETRIEVE DATA\n",
    "#clinical\n",
    "clinical_data = ccrcc.get_clinical()\n",
    "\n",
    "\n",
    "#rna\n",
    "rna_data = ccrcc.get_transcriptomics()\n",
    "\n",
    "\n",
    "#proteins\n",
    "protein_data = ccrcc.get_proteomics()\n",
    "#for easier manipulation\n",
    "protein_data.columns = protein_data.columns.get_level_values(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "id": "2cd4a99e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#CLEANUP\n",
    "#na mask. Filter out patients with NaN for tumor stage\n",
    "na_mask = clinical_data.loc[:, 'tumor_stage_pathological'].isna()\n",
    "#apply mask to data frames\n",
    "clinical_masked= clinical_data.loc[~na_mask, :]\n",
    "rna_masked= rna_data.loc[~na_mask, :]\n",
    "proteins_masked= protein_data.loc[~na_mask, :]\n",
    "\n",
    "\n",
    "#change 0s to NAs before removing\n",
    "rna_masked = rna_masked.replace(0, np.nan)\n",
    "#get rid of genes with no data\n",
    "gene_na_mask = rna_masked.isna().sum() == 0\n",
    "#cleanup rna df\n",
    "rna_masked = rna_masked.loc[:, gene_na_mask]\n",
    "\n",
    "\n",
    "#same thing for proteins. get rid of the ones w no data\n",
    "proteins_na_mask = proteins_masked.isna().sum() == 0\n",
    "#cleanup df\n",
    "proteins_masked = proteins_masked.loc[:, proteins_na_mask]\n",
    "\n",
    "\n",
    "#stage mask to divide patients into Stage I cohort and stage III cohort\n",
    "stg1_mask = clinical_masked.loc[:, 'tumor_stage_pathological'] == 'Stage I'\n",
    "stg3_mask= clinical_masked.loc[:, 'tumor_stage_pathological'] == 'Stage III'\n",
    "\n",
    "Y_data=clinical_masked.loc[: , 'tumor_stage_pathological']\n",
    "\n",
    "#logscale\n",
    "rna_masked= np.log2(rna_masked)\n",
    "\n",
    "\n",
    "#TOP 5 MOST DIFFERENTIALLY EXPRESSED RNA \n",
    "#get top 5 differentially expressed genes for RNA stage1\n",
    "s1_diff_rna = rna_masked.loc[stg1_mask, :].mean() - rna_masked.loc[stg3_mask, :].mean()\n",
    "s1_t5_rna = np.abs(s1_diff_rna).sort_values(ascending=False)[:5].index\n",
    "\n",
    "#TOP 5 MOST DIFFERENTIALLY EXPRESSED PROTEINS\n",
    "#Stage 1\n",
    "s1_diff_proteins= proteins_masked.loc[stg1_mask, :].mean()- proteins_masked.loc[stg3_mask, :].mean()\n",
    "s1_t5_proteins= np.abs(s1_diff_proteins).sort_values(ascending= False)[:5].index\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "id": "76db0c5f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/5x/ysrbhc4j6hldh0hl5gpqh_3m0000gn/T/ipykernel_37361/4117901636.py:12: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  s1s3_newdf.loc[: , s1s3_proteins[i]]= proteins_masked.loc[: , s1s3_proteins[i]]\n",
      "/var/folders/5x/ysrbhc4j6hldh0hl5gpqh_3m0000gn/T/ipykernel_37361/4117901636.py:12: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  s1s3_newdf.loc[: , s1s3_proteins[i]]= proteins_masked.loc[: , s1s3_proteins[i]]\n",
      "/var/folders/5x/ysrbhc4j6hldh0hl5gpqh_3m0000gn/T/ipykernel_37361/4117901636.py:12: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  s1s3_newdf.loc[: , s1s3_proteins[i]]= proteins_masked.loc[: , s1s3_proteins[i]]\n",
      "/var/folders/5x/ysrbhc4j6hldh0hl5gpqh_3m0000gn/T/ipykernel_37361/4117901636.py:12: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  s1s3_newdf.loc[: , s1s3_proteins[i]]= proteins_masked.loc[: , s1s3_proteins[i]]\n",
      "/var/folders/5x/ysrbhc4j6hldh0hl5gpqh_3m0000gn/T/ipykernel_37361/4117901636.py:12: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  s1s3_newdf.loc[: , s1s3_proteins[i]]= proteins_masked.loc[: , s1s3_proteins[i]]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Name</th>\n",
       "      <th>AJAP1</th>\n",
       "      <th>DPEP1</th>\n",
       "      <th>GALNT5</th>\n",
       "      <th>IL20RB</th>\n",
       "      <th>RYR2</th>\n",
       "      <th>FTL</th>\n",
       "      <th>HBZ</th>\n",
       "      <th>HBA2</th>\n",
       "      <th>CMA1</th>\n",
       "      <th>HBB</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C3L-00004</th>\n",
       "      <td>-2.519207</td>\n",
       "      <td>-1.239811</td>\n",
       "      <td>-5.281123</td>\n",
       "      <td>2.685910</td>\n",
       "      <td>-0.721420</td>\n",
       "      <td>-1.724339</td>\n",
       "      <td>-0.370098</td>\n",
       "      <td>-0.786307</td>\n",
       "      <td>-0.918273</td>\n",
       "      <td>-0.762287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00010</th>\n",
       "      <td>-0.069282</td>\n",
       "      <td>4.515644</td>\n",
       "      <td>-1.685391</td>\n",
       "      <td>0.394540</td>\n",
       "      <td>-2.134346</td>\n",
       "      <td>-0.363228</td>\n",
       "      <td>0.240576</td>\n",
       "      <td>-0.081263</td>\n",
       "      <td>0.747860</td>\n",
       "      <td>-0.034944</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00011</th>\n",
       "      <td>0.411676</td>\n",
       "      <td>-1.955016</td>\n",
       "      <td>-2.430473</td>\n",
       "      <td>5.329663</td>\n",
       "      <td>3.102518</td>\n",
       "      <td>-0.977364</td>\n",
       "      <td>-0.087641</td>\n",
       "      <td>-0.418663</td>\n",
       "      <td>-0.255054</td>\n",
       "      <td>-0.371128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00026</th>\n",
       "      <td>0.177869</td>\n",
       "      <td>4.635430</td>\n",
       "      <td>2.729468</td>\n",
       "      <td>1.139609</td>\n",
       "      <td>-1.565029</td>\n",
       "      <td>1.301930</td>\n",
       "      <td>0.981930</td>\n",
       "      <td>0.780756</td>\n",
       "      <td>0.049086</td>\n",
       "      <td>0.855097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00079</th>\n",
       "      <td>-1.161677</td>\n",
       "      <td>0.376214</td>\n",
       "      <td>-3.413559</td>\n",
       "      <td>0.441469</td>\n",
       "      <td>-0.421905</td>\n",
       "      <td>-1.496648</td>\n",
       "      <td>-0.441854</td>\n",
       "      <td>-1.014088</td>\n",
       "      <td>-0.620829</td>\n",
       "      <td>-0.939448</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3N-01646</th>\n",
       "      <td>-1.251963</td>\n",
       "      <td>-0.946572</td>\n",
       "      <td>-0.635367</td>\n",
       "      <td>7.399686</td>\n",
       "      <td>-0.645290</td>\n",
       "      <td>-1.817120</td>\n",
       "      <td>-0.724017</td>\n",
       "      <td>-1.180566</td>\n",
       "      <td>-0.000699</td>\n",
       "      <td>-1.127855</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3N-01648</th>\n",
       "      <td>-2.212814</td>\n",
       "      <td>2.734819</td>\n",
       "      <td>1.865296</td>\n",
       "      <td>0.789974</td>\n",
       "      <td>-1.221495</td>\n",
       "      <td>0.459593</td>\n",
       "      <td>-0.396638</td>\n",
       "      <td>-0.672383</td>\n",
       "      <td>-0.611409</td>\n",
       "      <td>-0.605164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3N-01649</th>\n",
       "      <td>-0.097150</td>\n",
       "      <td>3.354027</td>\n",
       "      <td>-1.442680</td>\n",
       "      <td>-0.220637</td>\n",
       "      <td>2.057367</td>\n",
       "      <td>-1.066512</td>\n",
       "      <td>-0.109630</td>\n",
       "      <td>-0.584516</td>\n",
       "      <td>0.322161</td>\n",
       "      <td>-0.527777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3N-01651</th>\n",
       "      <td>-3.769215</td>\n",
       "      <td>-0.554359</td>\n",
       "      <td>-1.765596</td>\n",
       "      <td>0.924430</td>\n",
       "      <td>3.111491</td>\n",
       "      <td>-1.895398</td>\n",
       "      <td>-0.096922</td>\n",
       "      <td>-0.649448</td>\n",
       "      <td>-0.302394</td>\n",
       "      <td>-0.571897</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3N-01808</th>\n",
       "      <td>-0.590156</td>\n",
       "      <td>4.861775</td>\n",
       "      <td>-2.466955</td>\n",
       "      <td>0.540075</td>\n",
       "      <td>-1.611685</td>\n",
       "      <td>1.895297</td>\n",
       "      <td>2.056017</td>\n",
       "      <td>1.485439</td>\n",
       "      <td>-1.125322</td>\n",
       "      <td>1.538612</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>110 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Name           AJAP1     DPEP1    GALNT5    IL20RB      RYR2       FTL  \\\n",
       "Patient_ID                                                               \n",
       "C3L-00004  -2.519207 -1.239811 -5.281123  2.685910 -0.721420 -1.724339   \n",
       "C3L-00010  -0.069282  4.515644 -1.685391  0.394540 -2.134346 -0.363228   \n",
       "C3L-00011   0.411676 -1.955016 -2.430473  5.329663  3.102518 -0.977364   \n",
       "C3L-00026   0.177869  4.635430  2.729468  1.139609 -1.565029  1.301930   \n",
       "C3L-00079  -1.161677  0.376214 -3.413559  0.441469 -0.421905 -1.496648   \n",
       "...              ...       ...       ...       ...       ...       ...   \n",
       "C3N-01646  -1.251963 -0.946572 -0.635367  7.399686 -0.645290 -1.817120   \n",
       "C3N-01648  -2.212814  2.734819  1.865296  0.789974 -1.221495  0.459593   \n",
       "C3N-01649  -0.097150  3.354027 -1.442680 -0.220637  2.057367 -1.066512   \n",
       "C3N-01651  -3.769215 -0.554359 -1.765596  0.924430  3.111491 -1.895398   \n",
       "C3N-01808  -0.590156  4.861775 -2.466955  0.540075 -1.611685  1.895297   \n",
       "\n",
       "Name             HBZ      HBA2      CMA1       HBB  \n",
       "Patient_ID                                          \n",
       "C3L-00004  -0.370098 -0.786307 -0.918273 -0.762287  \n",
       "C3L-00010   0.240576 -0.081263  0.747860 -0.034944  \n",
       "C3L-00011  -0.087641 -0.418663 -0.255054 -0.371128  \n",
       "C3L-00026   0.981930  0.780756  0.049086  0.855097  \n",
       "C3L-00079  -0.441854 -1.014088 -0.620829 -0.939448  \n",
       "...              ...       ...       ...       ...  \n",
       "C3N-01646  -0.724017 -1.180566 -0.000699 -1.127855  \n",
       "C3N-01648  -0.396638 -0.672383 -0.611409 -0.605164  \n",
       "C3N-01649  -0.109630 -0.584516  0.322161 -0.527777  \n",
       "C3N-01651  -0.096922 -0.649448 -0.302394 -0.571897  \n",
       "C3N-01808   2.056017  1.485439 -1.125322  1.538612  \n",
       "\n",
       "[110 rows x 10 columns]"
      ]
     },
     "execution_count": 223,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#MAKE NEW DF\n",
    "s1s3_rna= s1_t5_rna[:]\n",
    "s1s3_rna\n",
    "\n",
    "s1s3_proteins= s1_t5_proteins[:]\n",
    "\n",
    "#rows =patients and the columns =expression values of selected genes\n",
    "s1s3_newdf = rna_masked[s1s3_rna]\n",
    "\n",
    "#add protein data\n",
    "for i in range(5):\n",
    "    s1s3_newdf.loc[: , s1s3_proteins[i]]= proteins_masked.loc[: , s1s3_proteins[i]]\n",
    "    \n",
    "s1s3_newdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "id": "60ff7366",
   "metadata": {},
   "outputs": [],
   "source": [
    "#ENCODE AND SCALE\n",
    "scaler = StandardScaler()\n",
    "s1s3_newdf = scaler.fit_transform(s1s3_newdf)\n",
    "\n",
    "encoder= LabelEncoder()\n",
    "Y_data= encoder.fit_transform(Y_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "id": "3742e24b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Average accuracy for the given classifier:  0.49696969696969706\n",
      "Average accuracy for the given classifier:  0.3939393939393939\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/andreavelazquez/anaconda3/lib/python3.11/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.\n",
      "  warnings.warn(\n",
      "/Users/andreavelazquez/anaconda3/lib/python3.11/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.\n",
      "  warnings.warn(\n",
      "/Users/andreavelazquez/anaconda3/lib/python3.11/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.\n",
      "  warnings.warn(\n",
      "/Users/andreavelazquez/anaconda3/lib/python3.11/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.\n",
      "  warnings.warn(\n",
      "/Users/andreavelazquez/anaconda3/lib/python3.11/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.\n",
      "  warnings.warn(\n",
      "/Users/andreavelazquez/anaconda3/lib/python3.11/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.\n",
      "  warnings.warn(\n",
      "/Users/andreavelazquez/anaconda3/lib/python3.11/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.\n",
      "  warnings.warn(\n",
      "/Users/andreavelazquez/anaconda3/lib/python3.11/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Average accuracy for the given classifier:  0.5272727272727273\n",
      "Average accuracy for the given classifier:  0.5666666666666667\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/andreavelazquez/anaconda3/lib/python3.11/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.\n",
      "  warnings.warn(\n",
      "/Users/andreavelazquez/anaconda3/lib/python3.11/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "#SPLITS\n",
    "from sklearn.model_selection import train_test_split\n",
    "X_train, X_test, y_train, y_test = train_test_split(s1s3_newdf, Y_data, train_size=0.7)\n",
    "\n",
    "#ACCURACY: KNeighborsClassifier, DecisionTreeClassifier, and MLPClassifier, GaussianNB\n",
    "from sklearn.neighbors import KNeighborsClassifier # default number of neighbors looked at is 5\n",
    "from sklearn.tree import DecisionTreeClassifier\n",
    "from sklearn.neural_network import MLPClassifier\n",
    "from sklearn.naive_bayes import GaussianNB\n",
    "\n",
    "#KNeighborsClassifier\n",
    "def avg_acc_classifier(funct):\n",
    "    \n",
    "    #list to save results\n",
    "    results=[]\n",
    "    \n",
    "    for i in range(10):\n",
    "        #Split df into training and testing. Diff split each time\n",
    "        X_train, X_test, y_train, y_test = train_test_split(s1s3_newdf, Y_data, train_size=0.7)\n",
    "\n",
    "        #use classifier \n",
    "        classifier = funct\n",
    "        classifier.fit(X_train, y_train)\n",
    "        y_pred = classifier.predict(X_test)\n",
    "        accuracy = sum(y_pred == y_test) / len(y_test)\n",
    "        results.append(accuracy)\n",
    "    \n",
    "    avg= sum(results)/len(results)\n",
    "\n",
    "    print(\"Average accuracy for the given classifier: \", avg)\n",
    "    \n",
    "    return \n",
    "\n",
    "KNN= avg_acc_classifier(KNeighborsClassifier())\n",
    "DT= avg_acc_classifier(DecisionTreeClassifier())\n",
    "MLP= avg_acc_classifier(MLPClassifier())\n",
    "GNB= avg_acc_classifier(GaussianNB())\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
